cli/src/pcluster/models/cluster.py

```diff
@@ -510,69 +555,165 @@ def head_node_ip(self):
+
+            else:  # scheduler == "slurm"
+                stack_status = self.stack.status
+                if "IN_PROGRESS" in stack_status:
+                    raise ClusterActionError(f"Cannot stop compute fleet while stack is in {stack_status} status.")
+                if "FAILED" in stack_status:
+                    LOGGER.warning("Cluster stack is in %s status. This operation might fail.", stack_status)
 
-            try:
                 compute_fleet_status_manager = ComputeFleetStatusManager(self.name)
                 compute_fleet_status_manager.update_status(
                     ComputeFleetStatus.STOP_REQUESTED, ComputeFleetStatus.STOPPING, ComputeFleetStatus.STOPPED
                 )
-            except ComputeFleetStatusManager.ConditionalStatusUpdateFailed:
-                raise ClusterActionError(
-                    "Failed when stopping compute fleet due to a concurrent update of the status. "
-                    "Please retry the operation."
+        except ComputeFleetStatusManager.ConditionalStatusUpdateFailed:
+            raise ClusterActionError(
+                "Failed when stopping compute fleet due to a concurrent update of the status. "
+                "Please retry the operation."
+            )
+        except ClusterActionError as e:
+            raise e
+        except Exception as e:
+            raise ClusterActionError(f"Failed when stopping compute fleet with error: {str(e)}")
+
+    def update(
+        self,
+        cluster_config: dict,
+        suppress_validators: bool = False,
+        validation_failure_level: FailureLevel = FailureLevel.ERROR,
+        force: bool = False,
+    ):
+        """Update cluster."""
+        try:
+            # check cluster existence
+            if not AWSApi.instance().cfn.stack_exists(self.stack_name):
+                raise ClusterActionError(f"Cluster {self.name} does not exist")
+
+            if "IN_PROGRESS" in self.stack.status:
+                raise ClusterActionError(f"Cannot execute update while stack is in {self.stack.status} status.")
+
+            # validate target config
+            self._validate_and_parse_config(suppress_validators, validation_failure_level, cluster_config)
+
+            # verify changes
+            patch = ConfigPatch(cluster=self, base_config=self.source_config, target_config=cluster_config)
+            patch_allowed, update_failures = patch.check()
+            if not (patch_allowed or force):
+                raise ClusterActionError("Update failure", update_failures=update_failures)
+
+            # Retrieve bucket information
+            self.bucket = ClusterBucket(
+                name=self.stack.s3_bucket_name,
+                artifact_directory=self.stack.s3_artifact_directory,
+                remove_on_deletion=True,
+            )
+            self._add_version_tag()
+            self._upload_config()
+
+            # Create template if not provided by the user
```

Comment:
I can't figure out what we are doing here. Maybe some more comments could help?

Reply:
We're building the template with CDK, if not specified by the user with the DEV settings param.